app/upgrade_handler.go

```diff
@@ -97,3 +102,68 @@ func (app ShentuApp) setTmpUpgradeHandler() {
 		app.SetStoreLoader(upgradetypes.UpgradeStoreLoader(upgradeInfo.Height, &storeUpgrades))
 	}
 }
+
+func RunShieldMigration(app ShentuApp, ctx sdk.Context) {
+	sk := app.ShieldKeeper
+	bondDenom := sk.BondDenom(ctx)
+
+	moduleCoins := app.BankKeeper.GetAllBalances(ctx, app.AccountKeeper.GetModuleAccount(ctx, shieldtypes.ModuleName).GetAddress())
+
+	// remaining service fees
+	remainingServiceFees := sk.GetRemainingServiceFees(ctx)
+
+	// rewards
+	var rewards sdk.DecCoins
+	for _, provider := range sk.GetAllProviders(ctx) {
+		rewards = rewards.Add(provider.Rewards...)
+	}
+
+	totalInt, _ := remainingServiceFees.Add(rewards...).TruncateDecimal()
```

Comment:
Since there is check requiring the decimal part of the adding result is empty in shield's invariant, shall we make sure the decimal part is zero here?

Reply:
Hmm Yes I think that is the case. I'll add that in.